File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
    let edges = 0;
    const visited = Array(n).fill(0);
    const graph = Array.from({length: n}, () => Array(0));
    
    for (let i = 0 ; i < n ; i++) {
        for (let j = i + 1 ; j < n ; j++) {
            if (computers[i][j] === 1) {
                graph[i].push(j);
                graph[j].push(i);
            }
        }
    }
    
    function dfs(cur) {        
        visited[cur] = true;
        for (const next of graph[cur]) {
            if (!visited[next]) {
                dfs(next);
                edges += 1;
            }
        }
    }
    
    for (let i = 0 ; i < n ; i++) {
        if (!visited[i]) {
            dfs(i);                
        }
    }
    
    return n - edges;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    
    visited = [False for _ in range(n)]
    
    answer = 0

    for start in range(n):
        if not visited[start]:
            dfs(start, computers, visited)
            answer += 1
        
    return answer

def dfs(cur, computers, visited):
    
    visited[cur] = True
    
    for nxt in range(len(computers)):
        
        if cur != nxt and computers[cur][nxt] and not visited[nxt]:
            dfs(nxt, computers, visited)
